clients/drcachesim/tracer/func_trace.cpp

```diff
@@ -0,0 +1,172 @@
+/* **********************************************************
+ * Copyright (c) 2016-2018 Google, Inc.  All rights reserved.
+ * **********************************************************/
+
+/*
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * * Redistributions of source code must retain the above copyright notice,
+ *   this list of conditions and the following disclaimer.
+ *
+ * * Redistributions in binary form must reproduce the above copyright notice,
+ *   this list of conditions and the following disclaimer in the documentation
+ *   and/or other materials provided with the distribution.
+ *
+ * * Neither the name of Google, Inc. nor the names of its contributors may be
+ *   used to endorse or promote products derived from this software without
+ *   specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL VMWARE, INC. OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+ * DAMAGE.
+ */
+
+/* func_trace.cpp: code for recording function traces */
+
+#include <vector>
+#include <string>
+#include "dr_api.h"
+#include "drsyms.h"
+#include "drwrap.h"
+#include "drmgr.h"
+#include "trace_entry.h"
+#include "../common/options.h"
+#include "func_trace.h"
+
+typedef struct {
+    std::string name;
+    int id;
+    int arg_num;
+} func_metadata_t;
+
+/* expected pattern "<function_name>:<function_id>:<arguments_num>" */
+#define PATTERN_SEPARATOR ":"
+
+
+static int func_trace_init_count;
+
+static func_trace_append_entry_t append_entry;
+static std::vector<func_metadata_t> funcs;
```

Comment:
Note, about using std::vector (and the std::string above) in a DR client (instead of something like drvector_t). Because in our use-case we are statically linked with the application, you have to be very careful not to break isolation. For example, if func_pre_hook() was calling funcs.push_back(), that could call the same malloc we are currently tracing, which is definitely something we don't want to do, and will lead to all kinds of trouble.

I think you're fine in this case (but @derekbruening sure knows better) -- you're only indexing in the vector when we are called from tracing code. Probably still worth leaving a short comment for the next reader.

Reply:
This is a very good point, I think very likely it will happen. I will update to use drvector_t and avoid to use std::string in the function hook.

Reply:
Done